File: advent-of-code/2015/02/02/solution.zig
const std = @import("std");

fn calculateRibbon(dimensions: []const u8) !u32 {
    var it = std.mem.splitScalar(u8, dimensions, 'x');
    var dims_array: [3]u32 = undefined;
    var i: usize = 0;

    while (it.next()) |part| {
        const num = try std.fmt.parseInt(u32, part, 10);
        dims_array[i] = num;
        i += 1;
    }

    // Sort dimensions
    std.mem.sort(u32, &dims_array, {}, comptime std.sort.asc(u32));

    // Smallest perimeter (using two smallest dimensions)
    const perimeter = 2 * (dims_array[0] + dims_array[1]);

    // Volume for bow
    const volume = dims_array[0] * dims_array[1] * dims_array[2];

    return perimeter + volume;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();

    // Test cases with assertions
    std.debug.assert(try calculateRibbon("2x3x4") == 34);
    std.debug.assert(try calculateRibbon("1x1x10") == 14);

    const input = try std.fs.cwd().readFileAlloc(gpa.allocator(), "../input.txt", 1024 * 1024);
    defer gpa.allocator().free(input);

    var total: u32 = 0;
    var lines = std.mem.splitScalar(u8, input, '\n');
    while (lines.next()) |line| {
        if (line.len > 0) {
            total += try calculateRibbon(line);
        }
    }
    std.debug.print("Zig: {d}\n", .{total});
}

File: advent-of-code/2015/02/02/solution.clj
(require '[clojure.string :as str])

(defn calculate-ribbon [dimensions]
  (let [dims (sort (map #(Integer/parseInt %) (str/split dimensions #"x")))
        ;; Smallest perimeter (using two smallest dimensions)
        perimeter (* 2 (+ (first dims) (second dims)))
        ;; Volume for bow
        volume (reduce * dims)]
    (+ perimeter volume)))

;; Test cases with assertions
(assert (= 34 (calculate-ribbon "2x3x4")))
(assert (= 14 (calculate-ribbon "1x1x10")))

;; Read input and calculate total
(let [script-dir (.getParent (java.io.File. *file*))
      input-path (str script-dir "/../input.txt")
      total (->> (slurp input-path)
                 str/split-lines
                 (filter #(not (str/blank? %)))
                 (map calculate-ribbon)
                 (reduce +))]
  (println "Clojure:" total))
